doc/Decoder.rst

```diff
@@ -0,0 +1,73 @@
+
+The use of an external scorer is fully optional. When an external scorer is not specified, DeepSpeech still uses a beam search decoding algorithm, but without any outside scoring.
+
+Currently, the DeepSpeech external scorer is implemented with `KenLM <https://kheafield.com/code/kenlm/>`_, plus some tooling to package the necessary files and metadata into a single ``.scorer`` package. The tooling lives in ``data/lm/``. The scripts included in ``data/lm/`` can be used and modified to build your own language model based on your particular use case or language.
+
+The scripts are geared towards replicating the language model files we release as part of `DeepSpeech model releases <https://github.com/mozilla/DeepSpeech/releases/latest>`_, but modifying them to use different datasets or language model construction parameters should be simple.
+
+
+Decoding modes
+^^^^^^^^^^^^^^
+
+DeepSpeech currently supports two modes of operation with significant differences at both training and decoding time.
+
+
+Default mode (alphabet based)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The default mode, which uses an alphabet file (specified with ``--alphabet_config_path`` at training and export time) to determine which labels (characters), and how many of them, to predict in the output layer. At decoding time, if using an external scorer, it MUST be word based and MUST be built using the same alphabet file used for training. Word based means the text corpus used to build the scorer should contain words separated by whitespace. For most western languages, this is the default and requires no special steps from the developer part when creating the scorer.
+
+
+UTF-8 mode
+^^^^^^^^^^
+
+In UTF-8 mode the model predicts UTF-8 bytes directly instead of letters from an alphabet file. This idea was proposed in the paper `Bytes Are All You Need <https://arxiv.org/abs/1811.09021>`_. This mode is enabled with the ``--utf8`` flag at training and export time. At training time, the alphabet file is not used. Instead, the model is forced to have 256 labels, with labels 0-254 corresponding to UTF-8 byte values 1-255, and label 255 is used for the CTC blank symbol. At decoding time, if using an external scorer, it MUST be built according to the instructions that follow.
+
+UTF-8 decoding can be useful for languages with very large alphabets, like Chinese Mandarin. It may also be useful for building multi-language models, or as a base for transfer learning. Currently these cases are untested and unsupported. Currently, UTF-8 mode makes assumptions that hold for Chinese Mandarin and may not hold for other languages.
```

Comment:
> "Chinese Mandarin" => "Mandarin written using Simplified Chinese"

For the reasoning see a similar comment later in the review.

Reply:
> Currently these cases are untested and unsupported. Currently, UTF-8 mode makes assumptions that hold for Chinese Mandarin and may not hold for other languages.

Should be changed to 

> Currently these cases are untested and unsupported. Also note that UTF-8 mode makes assumptions that hold for Mandarin written using Simplified Chinese and may not hold for other languages.

to avoid repetition of "Currently".